rfcs/20230210-sparsity.md

```diff
@@ -0,0 +1,373 @@
+* **singleton** : a variant of the compressed format,
+  for when coordinates have no siblings.
+
+The specification should be easy to extend for adding new level-formats.
+For example, a parameterized type **compressed(2:4)** could be used to
+define 2-out-of-4 structured sparsity that is well-supported by some
+vendors (viz. NVidia GPUs). These three level-formats have two properties:
+**non/unique** (are duplicates allowed at that level), and
+**un/ordered** (do coordinates appear sorted at that level). Both of
+which are optional, with the defaults being unique and ordered; but
+the encoding should allow specifying the non-unique and unordered
+settings. However, the dense format must always be ordered and
+unique; so the encoding should also be able to prohibit users from
+specifying invalid level-types like “unordered dense”.
+
+The encoding outlined above allows for a formal specification of most
+common sparse storage schemes (e.g., `COO`, `CSR`, `CSC`, `DCSR`,
+`DCSC`, `CSF`, `BCSR`, and `ELL`). Other extensions to the
+encoding, such as found in the paper by
+[Chou et al.](http://tensor-compiler.org/publications.html)
+are open to debate as well.
+
+The exact syntax for the encoding is still open for debate
+(it will be finalized and formalized in the final spec).
+We will give a concrete proposal as a separate RFC, since
+doing so gets into some technical details that are independent
+of the rest of the current RFC. Regardless of the particular
+syntax used for the encoding, the essence of the proposal
+is that a regular tensor type such as the following 10×20
+matrix type with 32-bit floating-point elements
+
+    tensor<10x20xf32>
+
+is made sparse by adding a single field
+
+    tensor<10x20xf32, #CSR>
+
+where the field is defined elsewhere to specify the encoding parts
+(1)-(2) described above. The `CSR` storage scheme is defined by
+the following encoding, where each dimension is mapped to a level
+in the same order:
+
+    #CSR = #sparse_tensor.encoding<
+      (i, j) -> (i : dense, j : compressed)
+    >
+
+The `CSC` storage scheme is very similar, but the mapping from
+dimensions to levels is a permutation:
+
+    #CSC = #sparse_tensor.encoding<
+      (i, j) -> (j : dense, i : compressed)
+    >
+
+Block sparsity with 2×3 blocks is expressed as follows:
+
+    #BCSR = #sparse_tensor.encoding<
+      (i, j) ->
+        ( i floordiv 2 : compressed
+        , j floordiv 3 : compressed
+        , i mod 2 : dense
+        , j mod 3 : dense
+        )
+     >
+
+For the block sparsity example, the dimension-expressions can
+be inferred automatically. But here’s what it would look like
+to specify them explicitly (just to give an idea of the syntax):
+
+    #BCSR = #sparse_tensor.encoding<
+      ( i = ib * 3 + ii
+      , j = jb * 2 + jj
+      ) ->
+      ( ib = i floordiv 2 : compressed
+      , jb = j floordiv 3 : compressed
+      , ii = i mod 2 : dense
+      , jj = j mod 3 : dense
+      )

```

Comment:
I really like this concept of expressing the affine mapping of physical (data-structure) indices to logical indices. I do think it's a bit unclear how this mapping will be used to generate good sparse code. For example, how would this mapping enable the compiler to generate an element-wise CSR * BCSR that properly coiterates the `j` dimension?

Reply:
First, thanks for all your feedback! Much appreciated. 

You bring up a good point on the tension between keeping the sparse tensor types very general and the ability of a sparse compiler to ultimately map this to efficient code. I was trying to address this in my " healthy sparse ecosystem" comment below, but should probably spell this out a bit more clear.  The general direction I want to go is that sparse compilers should probably map any combination of sparsity to functional code, even if some of these are not very efficient (I am a bit weary on just giving up for inefficient case). Finding what combination works well is then a "healthy" cooperation between ML developer, the framework that generates StableHLO, and sparse the compiler.